src/lambda/wizards/samDeployWizard.ts

```diff
@@ -0,0 +1,228 @@
+/*!
+ * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+
+import * as nls from 'vscode-nls'
+const localize = nls.loadMessageBundle()
+
+import * as path from 'path'
+import * as vscode from 'vscode'
+import { detectLocalTemplates } from '../local/detectLocalTemplates'
+import { MultiStepWizard, WizardStep } from './multiStepWizard'
+
+export interface SamDeployWizardArgs {
+    template: vscode.Uri
+    s3Bucket: string
+    stackName: string
+}
+
+type WorkspaceFolderPickUri = Pick<vscode.WorkspaceFolder, 'uri'>
+
+export interface SamDeployWizardContext {
+    readonly onDetectLocalTemplates: typeof detectLocalTemplates
+
+    readonly workspaceFolders: WorkspaceFolderPickUri[] | undefined
+
+    showInputBox: typeof vscode.window.showInputBox
+
+    showQuickPick: typeof vscode.window.showQuickPick
+}
+
+class DefaultSamDeployWizardContext implements SamDeployWizardContext {
+    public readonly onDetectLocalTemplates = detectLocalTemplates
+
+    public readonly showInputBox = vscode.window.showInputBox
+
+    public readonly showQuickPick = vscode.window.showQuickPick
+
+    public get workspaceFolders(): WorkspaceFolderPickUri[] | undefined {
+        return vscode.workspace.workspaceFolders
+    }
+}
+
+export class SamDeployWizard extends MultiStepWizard<SamDeployWizardArgs> {
```

Comment:
Consider using `Partial` to make code more DRY instead of defining all params twice (once required once optional). You can define class, `DeployWizard`, with required properties and use `Partial` to define type for constructor. Perhaps something like:
```constructor(params: Partial <SamDeployWizard> = {})```
or even perhaps:
```constructor(params: Partial <SamDeployWizard> & Partial <SamDeployWizardContext> = {})```